### PR TITLE
initial PoseidonTree draft

### LINKED ISSUE #25: Create general datastructure PoseidonTree, for storing general zk-friendly datastructures
In order to generalize over different tree types, we need a structure that generalize over different entities that we want to store, and the length of the proof branches.

The goal is to be able to write `PoseidonTree<Note>`, `PoseidonTree<Nullifier>` and specify the branch depth.

`PoseidonTree<T>` needs `T` to implement the trait `AsScalar`, which calculates a scalar based on the type.

### PR DESCRIPTION
**Pull request by krl**
Closes #25

### CODE REVIEW

File: src/tree/mod.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+            Ok(extend_scalar(root, self.branch_depth))
+        }
+    }
+
+    /// Returns a poseidon branch pointing at the specific index
+    ///
+    /// This includes padding the value to the correct branch length equivalent
+    pub fn poseidon_branch(
+        &self,
+        idx: u64,
```
Comment (CPerezz) [inline src/tree/mod.rs:78]: Shouldn't it ask for the height? So it can know how much should it pad?

**Review by CPerezz — APPROVED** [review]
I like that approach a lot. We should continue on that direction for our development!!
@ZER0 any opinions?

File: src/tree/mod.rs (hunk 1 above)
Comment (CPerezz) [reply src/tree/mod.rs:78]: Nevermind, you have it on the PoseidonTree stored. My bad

Comment (krl) [reply src/tree/mod.rs:78]: It pads based on the self.branch_depth, no reason to pass it here.

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "poseidon252"
-version = "0.4.0"
+version = "0.5.0"
 authors = [
   "zer0 <matteo@dusk.network>", "vlopes11 <victor@dusk.network>", "CPerezz <carlos@dusk.network>"
```
Comment (CPerezz) [inline Cargo.toml]: You should add yourself to authors @krl

File: src/tree/mod.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    ) -> io::Result<Option<Branch<NStack<T, PoseidonAnnotation, H>, H>>> {
+        self.inner.get(idx)
+    }
+
+    /// Get a mutable branch reference to the element at index `idx`, if any
+    pub fn get_mut(
+        &mut self,
+        idx: u64,
```
Comment (CPerezz) [inline src/tree/mod.rs:105]: Should we really force the index to be a u64??

**Review by CPerezz — APPROVED** [review]
LGTM but with a few questions

**Comment by CPerezz** [comment]
I have another question @krl .
Do we have any way to search leaves on the tree by value instead of by key?

**Comment by krl** [comment]
I was asked to leave a comment about the approach taken here.

This tree structure is a method of unifying all zk-proof capable datastructures we are using. The only requirement of the types stored in this tree as leaves, is that a StorageScalar can be computed from a reference to them. `for<'any> &'any T: Into<StorageScalar>`

The tree hashing is handled automagically by `kelvin`.

Additionally, when creating the structure, you define it's proof branch length, which is the length of the PoseidonBranch merkle proof, that can be used to do zero knowledge inclusion proofs.

If the tree is not actually that deep, the branch is padded with extra root nodes with only one child, until the specific length is reahed.

The `root()` call on the tree, that returns the scalar root hash, also does the same padding, to provide a "virtual root" at the specified height.

**Pull request merged**